grid/src/Cabana_Grid_Splines.hpp

```diff
@@ -451,6 +451,586 @@ struct Spline<3>
+      grid space.
+      \param values Basis values at the knots. Ordered from lowest to highest in
+      terms of knot location.
+    */
+    template <typename Scalar>
+    KOKKOS_INLINE_FUNCTION static void value( const Scalar x0,
+                                              Scalar values[num_knot] )
+    {
+        // Constants
+        Scalar denom_2 = 1.0 / 3840.0;
+        Scalar denom_1 = denom_2 * 2.0;
+
+        // Knot at i - 2
+        Scalar xn = x0 - static_cast<int>( x0 );
+        xn -= 0.5;
+        values[0] =
+            ( 1.0 +
+              xn * ( -10.0 +
+                     xn * ( 40.0 +
+                            xn * ( -80.0 +
+                                   xn * ( 80.0 + xn * ( -32.0 ) ) ) ) ) ) *
+            denom_2;
+
+        // Knot at i - 1
+        values[1] =
+            ( 237.0 +
+              xn * ( -750.0 +
+                     xn * ( 840.0 +
+                            xn * ( -240.0 +
+                                   xn * ( -240.0 + xn * ( 160.0 ) ) ) ) ) ) *
+            denom_2;
+
+        // Knot at i
+        values[2] =
+            ( 841.0 +
+              xn * ( -770.0 +
+                     xn * ( -440.0 +
+                            xn * ( 560.0 +
+                                   xn * ( 80.0 + xn * ( -160.0 ) ) ) ) ) ) *
+            denom_1;
+
+        // Knot at i + 1
+        values[3] =
+            ( 841.0 +
+              xn * ( 770.0 +
+                     xn * ( -440.0 +
+                            xn * ( -560.0 +
+                                   xn * ( 80.0 + xn * ( 160.0 ) ) ) ) ) ) *
+            denom_1;
+
+        // Knot at i + 2
+        values[4] =
+            ( 237.0 +
+              xn * ( 750.0 +
+                     xn * ( 840.0 +
+                            xn * ( 240.0 +
+                                   xn * ( -240.0 + xn * ( -160.0 ) ) ) ) ) ) *
+            denom_2;
+
+        // Knot at i + 3
+        values[5] =
+            ( 1.0 + xn * ( 10.0 +
+                           xn * ( 40.0 +
+                                  xn * ( 80.0 +
+                                         xn * ( 80.0 + xn * ( 32.0 ) ) ) ) ) ) *
+            denom_2;
+    }
+
+    /*!
+      \brief Calculate the value of the gradient of the spline in the physical
+      frame.
+      \param x0 The coordinate at which to evaluate the spline in the logical
+      grid space.
+      \param rdx The inverse of the physical distance between grid locations.
+      \param gradients Basis gradient values at the knots in the physical frame.
+      Ordered from lowest to highest in terms of knot location.
+    */
+    template <typename Scalar>
+    KOKKOS_INLINE_FUNCTION static void
+    gradient( const Scalar x0, const Scalar rdx, Scalar gradients[num_knot] )
+    {
+        // Constants
+        Scalar denom_2 = 1.0 / 384.0;
+        Scalar denom_1 = denom_2 * 2.0;
+
+        // Knot at i - 2
+        Scalar xn = x0 - static_cast<int>( x0 );
```

Comment:
Same as for quartic values. Precompute the powers of `xn` and change formatting to improve readability of the calls for `values[n]=`